File: src/engine/validationResult.ts
/**
 * Shared helper for building the validation portion of a RequestResult.
 * Used by both requestExecution.ts (load testing) and graphRunnerHelpers.ts (workflow engine).
 */
import type { FailureDetail, ValidationConfig, Assertion, TransportType } from '../shared/types';
import { validate, evaluateAssertions } from './validator';

export interface ValidationInput {
  httpStatus: number;
  responseTimeMs: number;
  responseHeaders: Record<string, string>;
  responseBody: string;
  responseObj: unknown;
  errorMessage?: string;
  validation: ValidationConfig;
  assertions: Assertion[];
  /** Populated by kafkaExecution for `kafkaField` assertion evaluation. */
  kafkaContext?: {
    key?: string;
    offset?: number;
    partition?: number;
    topic?: string;
  };
  /** Populated by WS execution for `wsField` / `wsNumericField` assertion evaluation. */
  wsContext?: {
    connectionId?: string;
    frameType?: 'text' | 'binary';
    protocol?: string;
    messageSize?: number;
    latencyMs?: number;
    url?: string;
  };
  /**
   * Transport type. Absent or `'http'` means standard HTTP request.
   * When non-HTTP, the `(http)` failure check (status 0 / 4xx+) is skipped
   * because HTTP status codes are synthetic for non-HTTP transports.
   */
  transportType?: TransportType;
}

export interface ValidationOutput {
  failureDetails: FailureDetail[];
  passed: boolean;
  errorMessage?: string;
}

/**
 * Evaluate assertions + JSON validation and determine pass/fail.
 * Centralises the logic previously duplicated in requestExecution and graphRunnerHelpers.
 */
export function buildValidationResult(input: ValidationInput): ValidationOutput {
  const { httpStatus, responseTimeMs, responseHeaders, responseObj, validation, assertions } = input;
  const { errorMessage } = input;
  const isHttpTransport = !input.transportType || input.transportType === 'http';

  const { failures: assertionFailures, statusAsserted } = assertions.length > 0
    ? evaluateAssertions(assertions, {
        httpStatus, responseTimeMs, responseHeaders,
        responseBody: responseObj, rawBody: input.responseBody,
        kafkaContext: input.kafkaContext, wsContext: input.wsContext,
      })
    : { failures: [], statusAsserted: false };

  const httpOk = httpStatus > 0 && httpStatus < 400;
  const statusOk = statusAsserted
    ? assertionFailures.every(f => f.path !== '(status)')
    : httpOk;

  const jsonFailures = validation.mode !== 'none' && statusOk
    ? validate(validation, responseObj)
    : [];

  let failureDetails = [...assertionFailures, ...jsonFailures];

  // HTTP failure: status >= 400 or network error (status 0), AND no status assertion
  // explicitly accepted it. Only applies to HTTP transports — for Kafka and WS,
  // httpStatus is synthetic (200 on success, 0 on failure) and should not generate
  // an `(http)` failure detail.
  if (isHttpTransport) {
    const httpFailed = !statusAsserted && (httpStatus >= 400 || httpStatus === 0);
    if (httpFailed) {
      const actual = errorMessage
        || (httpStatus === 0 ? 'network error' : `HTTP ${httpStatus}`);
      failureDetails = [{ path: '(http)', expected: '2xx', actual }, ...assertionFailures];
    }
  }

  const networkError = isHttpTransport && httpStatus === 0 && !statusAsserted;
  const nonHttpError = !isHttpTransport && (!!errorMessage || httpStatus === 0);
  const passed = !networkError && !nonHttpError && failureDetails.length === 0;

  return { failureDetails, passed, errorMessage };
}

File: src-tauri/src/validation_result.rs
use std::collections::HashMap;

use serde_json::Value;

use crate::assertion_evaluator::evaluate_assertions;
use crate::json_validator::validate;
use crate::validation_types::{
    Assertion, AssertionContext, FailureDetail, ValidationConfig, ValidationMode, ValidationOutput,
};

/// Port of `buildValidationResult()` from `validationResult.ts` (64 lines).
///
/// Evaluates assertions + JSON validation and determines pass/fail.
/// Replicates the exact combination logic from the JS implementation.
#[allow(clippy::too_many_arguments)]
pub fn build_validation_result(
    http_status: u16,
    response_time_ms: f64,
    response_headers: &HashMap<String, String>,
    response_body: &str,
    response_obj: &Value,
    error_message: Option<&str>,
    validation: &ValidationConfig,
    assertions: &[Assertion],
) -> ValidationOutput {
    // 1. Evaluate assertions if any
    let (assertion_failures, status_asserted) = if !assertions.is_empty() {
        let ctx = AssertionContext {
            http_status,
            response_time_ms,
            response_headers,
            response_body: response_obj,
            raw_body: response_body,
        };
        let result = evaluate_assertions(assertions, &ctx);
        (result.failures, result.status_asserted)
    } else {
        (vec![], false)
    };

    // 2. http_ok: status > 0 and < 400
    let http_ok = http_status > 0 && http_status < 400;

    // 3. status_ok: if status was explicitly asserted, check whether the assertion passed
    let status_ok = if status_asserted {
        assertion_failures
            .iter()
            .all(|f| f.path != "(status)")
    } else {
        http_ok
    };

    // 4. JSON validation: only when mode != none AND status_ok
    let json_failures = if validation.mode != ValidationMode::None && status_ok {
        validate(validation, response_obj)
    } else {
        vec![]
    };

    // 5. Merge: assertion_failures first, then json_failures
    let mut failure_details: Vec<FailureDetail> =
        [assertion_failures.clone(), json_failures].concat();

    // 6. HTTP failure overlay: when !status_asserted && (status >= 400 || status == 0)
    let http_failed = !status_asserted && (http_status >= 400 || http_status == 0);
    if http_failed {
        let actual = if let Some(msg) = error_message {
            if !msg.is_empty() {
                msg.to_string()
            } else if http_status == 0 {
                "network error".to_string()
            } else {
                format!("HTTP {}", http_status)
            }
        } else if http_status == 0 {
            "network error".to_string()
        } else {
            format!("HTTP {}", http_status)
        };

        let http_failure = FailureDetail {
            path: "(http)".to_string(),
            expected: "2xx".to_string(),
            actual,
        };
        // DROP json_failures — only keep [(http_failure), ...assertion_failures]
        failure_details = std::iter::once(http_failure)
            .chain(assertion_failures)
            .collect();
    }

    // 7-8. network_error and passed
    let network_error = http_status == 0 && !status_asserted;
    let passed = !network_error && failure_details.is_empty();

    ValidationOutput {
        failure_details,
        passed,
        error_message: error_message.map(|s| s.to_string()),
    }
}
